subsys/fs/ext2/ext2_impl.h

```diff
@@ -0,0 +1,342 @@
+/*
+ * Copyright (c) 2023 Antmicro <www.antmicro.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __EXT2_IMPL_H__
+#define __EXT2_IMPL_H__
+
+#include <zephyr/fs/fs.h>
+#include <zephyr/fs/ext2.h>
+
+#include "ext2_struct.h"
+
+void error_beharior(struct ext2_data *fs, const char *msg);
+
+/* Memory allocation for ext2 implementation */
+void *ext2_heap_alloc(size_t size);
+void  ext2_heap_free(void *ptr);
+
+/* Initialization of disk storage. */
+int disk_access_init_backend(struct ext2_data *fs, const void *storage_dev, int flags);
+
+/**
+ * @brief Get block from the disk.
+ */
+struct block *get_block(struct ext2_data *fs, uint32_t block);
+
+struct block *get_empty_block(struct ext2_data *fs);
+
+/**
+ * @brief Free the block structure.
+ */
+void drop_block(struct ext2_data *fs, struct block *b);
+
+/**
+ * @brief Write block to the disk.
+ *
+ * NOTICE: to ensure that all writes has ended the sync of disk must be triggered
+ * (fs::sync function).
```

Comment:
This is not how things are with the fatfs or lfs file systems, want @de-nordic to comment on if this is acceptable. This explains why whenever I wrote something on the device then opened it on the PC it just showed as empty and corrupt filesystem.

Reply:
So basically we have to ensure that user umounts file system when card is to be removed?
The sync is internal requirement, that is also handled by ext2 implementation? It is not that user has to call fs_sync() on file?
As far as user does not have to really sync after bunch of writes, this should be OK. Unfortunately sudden reset or shutdown with no umount will probably damage the system.

Reply:
What I meant here is that not every call to `ext2_write_block` ensures that changes are on disk. It only writes data to memory in a cache managed by `disk_access` implementation. I’ve added this comment for people reading my code, to point out that it isn’t enough to just use this function and be sure that everything is in the storage.

From the user point of view the situation is as you described. Sync is done during unmount and users don’t have to take any additional action. And of course, when the card is unplugged without unmounting (or system crashes) it might damage the file system.

Reply:
Had a talk with @de-nordic on this and we do not find this acceptable, file data needs to be written out when the fs_write functions are called to work as it does with other file systems and to prevent data loss, a user having to umount to write changes out is cumbersome, error prone, and basically breaks every in-tree (and probably all user applications) application if it was switched to using this file system.

Reply:
Okay. I will add the code that will ensure that everything is written to storage device at each write.

Reply:
I think that now the function`fs_sync` will not be needed at all (hence everything is flushed already). Am I right?

Reply:
Will have a look in a bit, it is OK to have uncommitted data in RAM whilst a file is open and fs_write has been used but it must be written when fs_close is used (i.e. function does not return until it has been written out) - that's fine, but not OK to withhold writing until fs_unmount is used. In the case where file is still open in write mode, fs_sync can be used to write the contents still in RAM out to the filesystem.

Reply:
> new data can remain volatile after fs_write is used but must be written to the disk when fs_close is used, it cannot stay volatile until fs_unmount is used - that is wrong

Okay, now I understand what you meant. I will add sync to the `fs_close` function and everything should be as you described.